azext_iot/common/base_discovery.py

```diff
@@ -0,0 +1,345 @@
+        resource types that require unique names within the subscription.
+
+        If the policy name is not provided, will look through all policies for the given
+        resource and return the first usable policy (the first policy that the IoT
+        extension can use).
+
+        Raises CLIError if no resource is found.
+
+        :param resource_name: Resource Name
+        :type resource_name: str
+        :param rg: Resource Group
+        :type rg: str
+
+        :keyword str login: Connection string for the target resource
+        :keyword str key_type: Key type to use in connection string construction
+        :keyword auth_type: Authentication Type for the Dataplane
+        :paramtype auth_type: AuthenticationTypeDataplane
+        :keyword str policy_name: Policy name to use
+
+        :return: Resource
+        :rtype: dict representing self.resource_type
+        """
+        cstring = kwargs.get("login")
+        if cstring:
+            return self.get_target_by_cstring(connection_string=cstring)
+
+        resource_group_name = resource_group_name or kwargs.get("rg")
+        target = self.find_resource(resource_name=resource_name, rg=resource_group_name)
+
+        key_type = kwargs.get("key_type", "primary")
+
+        # Azure AD auth path
+        auth_type = kwargs.get("auth_type", AuthenticationTypeDataplane.key.value)
+        if auth_type == AuthenticationTypeDataplane.login.value:
+            logger.info("Using AAD access token for %s interaction.", self.resource_type)
+            policy = SimpleNamespace()
+            policy.key_name = AuthenticationTypeDataplane.login.value
+            policy.primary_key = AuthenticationTypeDataplane.login.value
+            policy.secondary_key = AuthenticationTypeDataplane.login.value
+
+            return self._build_target(
+                resource=target,
+                policy=policy,
+                key_type="primary",
+                **kwargs
+            )
+
+        policy_name = kwargs.get("policy_name", "auto")
+        rg = target.additional_properties.get("resourcegroup")
+
+        target_policy = self.find_policy(
+            resource_name=target.name, rg=rg, policy_name=policy_name,
+        )
+
+        return self._build_target(
+            resource=target,
+            policy=target_policy,
+            key_type=key_type,
+            **kwargs
+        )
+
+    def get_targets(self, resource_group_name: str = None, **kwargs) -> List[Dict[str, str]]:
+        """
+        Returns a list of targets (dicts representing a resource's connection string parts)
+        that are usable by the extension within the subscription (and resource group if
+        provided).
+
+        :param rg: Resource Group
+        :type rg: str
+
+        :return: Resources
+        :rtype: list[dict]
+        """
+        targets = []
+        resources = self.get_resources(rg=resource_group_name)
+        if resources:
+            print()
```

Comment:
print()?

Reply:
Oh wow I did not event notice that... Thanks